src/ui/swipe.ts

```diff
@@ -0,0 +1,164 @@
+ * element.
+ *
+ *
+ * Usage:
+ * ```ts
+ * const swiper = new Swipe(myElement);
+ *
+ * swiper.on(Swipe.Events.UP, ()=> {
+ *   // swipe up.
+ * })
+ * swiper.on(Swipe.Events.LEFT, ()=> {
+ *   // swipe left
+ * })
+ *
+ * // Later
+ * swiper.dispose();
+ * ```
+ *
+ *
+ *
+ * Events are also fired on the element itself.
+ *
+ * ```ts
+ * const swiper = new Swipe(myElement);
+ * myElement.addEventListener(Swipe.Events.UP, ()=> {
+ *   // swipe up.
+ * })
+ * ```
+ */
+export class Swipe implements EventDispatcher {
+  private rootElement: HTMLElement;
+  private eventManager: EventManager;
+  private watcher: DomWatcher;
+  private x: Number;
+  private y: Number;
+  private isSwiping = false;
+  private allowSwipe = false;
+  private debounceDuration = 300;
+
+  constructor(rootElement: HTMLElement) {
+    this.rootElement = rootElement;
+    this.eventManager = new EventManager();
+    this.watcher = new DomWatcher();
+
+    const isTouchSupported = 'ontouchstart' in document.documentElement;
+
+    // Add listeners
+    if (isTouchSupported) {
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchstart',
+        eventOptions: {passive: true},
+        callback: this.onPointerDown.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchend',
+        eventOptions: {passive: true},
+        callback: this.onPointerUp.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'touchmove',
+        eventOptions: {passive: true},
+        callback: this.onPointerMove.bind(this),
+      });
+    } else {
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mousedown',
+        eventOptions: {passive: true},
+        callback: this.onPointerDown.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mouseup',
+        eventOptions: {passive: true},
+        callback: this.onPointerUp.bind(this),
+      });
+      this.watcher.add({
+        element: this.rootElement as HTMLElement,
+        on: 'mousemove',
+        eventOptions: {passive: true},
+        callback: this.onPointerMove.bind(this),
+      });
+    }
+  }
+
+  static Events = SwipeEvent;
+
+  private onPointerDown(e: TouchEvent) {
+    this.x = (e.touches && e.touches[0].clientX) || e['x'];
+    this.y = (e.touches && e.touches[0].clientY) || e['y'];
+    this.allowSwipe = true;
+  }
+
+  private onPointerUp() {
+    this.allowSwipe = false;
+  }
+
+  private onPointerMove(e: TouchEvent) {
+    if (!this.x || !this.y || this.isSwiping || !this.allowSwipe) {
+      return;
+    }
+
+    this.isSwiping = true;
+
+    const currentX = (e.touches && e.touches[0].clientX) || e['x'];
+    const currentY = (e.touches && e.touches[0].clientY) || e['y'];
+
+    const diffX = currentX - +this.x;
```

Comment:
Per style guide, use `Number(this.x)` instead.

(Although looking at this more closely, why is the number conversion here even necessary? In what cases would this not be a number?)

Reply:
Removed.  (it's not needed - I ported this and missed that)